[COBRA-3447] Consolidate apps & spacesapps tables

sbeck14 opened the pull request:
- Add V2 schema endpoints
  - `server/v2.go`
- Fields added to `spacesapps` table
- Add `v2.deployments` view
  - `apps` table no longer necessary
  - all fields pulled to view from `spacesapps` table
  - `appid` pulled from controller-api (if available) via fdw
    - temporary - will be removed once migration to v2 is complete
- Added `deployment` package
  - Functions that use the `v2.deployments` view
  - Actions on `deployments` i.e. deploy, remove, scale
- Added tests for `deployment` package

Still to do in the future (when ready to complete the migration):
- Look into combining `AddDeploymentV2` and `DeploymentV2`
  - One of these adds to the database, the other actually deploys to k8s
  - Any reason we can't deploy immediately when we add to the database?
- Write tests for v2 one-off deployments
  - Now a parameter in the payload of `PUT /v2beta1/space/:space/deployment:deployment/deploy`
- Documentation
- Go package management
  - Move `GET /v2beta1/apps` and `GET /v2beta1/app/:appid` to the apps package?
  - Move `DELETE /v2beta1/space/:space` to space package?

sbeck14 commented:
(Temporary) required environment variables:
- `V2_TEMP_CONTROLLER_API_DATABASE_HOST`
- `V2_TEMP_CONTROLLER_API_DATABASE_NAME`
- `V2_TEMP_CONTROLLER_API_DATABASE_USERNAME`
- `V2_TEMP_CONTROLLER_API_DATABASE_PASSWORD`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/akkeris/region-api?pullRequest=188) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/akkeris/region-api?pullRequest=188) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: sbeck14<br/>:x: Sam Beckett<br/><hr/>**Sam Beckett** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/akkeris/region-api?pullRequest=188) it.</sub>

sbeck14 commented:
I removed an old email of mine and I think that's why this ^ is happening. No way to put it back unfortunately

- Added `ENABLE_V2_ENDPOINTS` env var to control whether or not the V2 endpoints are enabled and tested

trevorlinton commented on deployment/deployment.go:
```diff
@@ -0,0 +1,1224 @@
+
+	_, err = db.Exec("DELETE from v2.deployments where space=$1 and name=$2", space, name)
+	if err != nil {
+		return 500, err
```
Can we use `http.Status*` codes instead of numbers?

sbeck14 replied:
Fixed!

trevorlinton commented on deployment/deployment.go:
```diff
@@ -0,0 +1,1224 @@
+package deployment
```
Within other portions of the region api we've distinguished (using separate files), what is a http end point and what is a library call, this would make it easier to know when modifying files what would potentially break a contract in addition to isolating dependencies.  See router/http.go as an example. This is optional, but wanted to give you an FYI.

sbeck14 replied:
That makes a ton of sense, but I don't think the space and app packages do that so I must have missed it. Will refactor a bit to match that pattern

trevorlinton commented on deployment/deployment.go:
```diff
@@ -0,0 +1,1224 @@
+		return
+	}
+
+	r.JSON(http.StatusCreated, structs.Messagespec{Status: http.StatusCreated, Message: "Deployment record created"})
```
I'm trying to deprecate the use of a `structs.Messagespec` as it has very little value.  Can we consider returning back the deployment object?

trevorlinton commented on deployment/deployment.go:
```diff
@@ -0,0 +1,1224 @@
+}
+
+// DeploymentV2Handler - HTTP handler for DeploymentV2
+func DeploymentV2Handler(db *sql.DB, payload structs.DeploySpecV2, berr binding.Errors, r render.Render) {
```
This function is kind of a beast, is there a way we can reduce its size and break it down into smaller functions that we can possibly unit test?

sbeck14 replied:
I spun the service creation (and filters) off into its own function which cut it down significantly. There might be more we can do here to reduce size as well

trevorlinton commented on v2_schema_migration.sql:
```diff
@@ -0,0 +1,163 @@
+			execute procedure v2_deployments_update_row();
+
+END
+$func$ LANGUAGE plpgsql;
```
Very good use of SQL to manage data 👍
